### PR TITLE
[Client] Allow Client{Object,Actor}Ref to accept a future.

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
This is in preparation for #18298. No code path actually creates refs with futures yet in this PR.

Another change is to check the number of returns and function signatures of tasks and methods on client side.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/client/worker.py [hunk 1]
```diff
@@ -281,7 +282,7 @@ def put(self, vals, *, client_ref_id: bytes = None):
             out = out[0]
         return out
 
-    def _put(self, val, *, client_ref_id: bytes = None):
+    def _put(self, val, client_ref_id: bytes):
```
Comment (ijrsvt) [inline python/ray/util/client/worker.py:285]: Any reason for dropping the `*` here?

**Review by ijrsvt — APPROVED** [review]
Overall LGTM!

File: python/ray/includes/object_ref.pxi [hunk 2]
```diff
@@ -134,24 +135,28 @@ cdef class ObjectRef(BaseID):
...
-    def _on_completed(self, py_callback: Callable[[Any], None]):
+    def _on_completed(self, set_future: Callable[[Any], None]):
         """Register a callback that will be called after Object is ready.
         If the ObjectRef is already ready, the callback will be called soon.
         The callback should take the result as the only argument. The result
         can be an exception object in case of task error.
         """
         core_worker = ray.worker.global_worker.core_worker
-        core_worker.set_get_async_callback(self, py_callback)
+        core_worker.set_get_async_callback(self, set_future)
```
[the remark this answers is not in the document]
Comment (mwtian) [reply python/ray/includes/object_ref.pxi:145]: Done, good point.

File: python/ray/util/client/worker.py (hunk 1 above)
Comment (mwtian) [reply python/ray/util/client/worker.py:285]: The `client_ref_id` arg is converted from optional keyword to become positional and always required. IIUC `*` becomes incorrect.

**Comment by mwtian** [comment]
> Overall LGTM!

Thanks for taking a look!

File: python/ray/util/client/server/dataservicer.py [hunk 3]
```diff
@@ -206,7 +205,8 @@ def Datapath(self, request_iterator, context):
                 # Connection isn't recoverable, skip cleanup
                 cleanup_requested = True
         finally:
-            logger.debug(f"Lost data connection from client {client_id}")
+            logger.debug(f"Stream is broken with client {client_id}")
+            self.basic_service.release_all(client_id)
```
Comment (ckw017) [inline python/ray/util/client/server/dataservicer.py:209]: the release_all here can be removed (it now happens farther down at what is currently line 241, after the grace period has passed). Guessing this was a remnant of resolving merge conflicts

Comment (mwtian) [reply python/ray/util/client/server/dataservicer.py]: Removed, thanks!

**Review by ckw017 — APPROVED** [review]
Looks good to me, let me know if you run into any weird problems with `test_client_reconnect` that might have popped up from the merge conf

**Comment by mwtian** [comment]
> Looks good to me, let me know if you run into any weird problems with `test_client_reconnect` that might have popped up from the merge conf

Sounds good.

**Comment by mwtian** [comment]
Failure in `//python/ray/data:tests/test_dataset` should be unrelated.

**Pull request merged**